### PR TITLE
[SPARK-39320][SQL] Support aggregate function `MEDIAN`

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Many mainstream database supports aggregate function `MEDIAN`.

**Syntax：**
Aggregate function
`MEDIAN( <expr> )`

Window function
`MEDIAN( <expr> ) OVER ( [ PARTITION BY <expr2> ] )`
**Arguments：**
expr: The expression must evaluate to a numeric data type (INTEGER, FLOAT, DECIMAL, or equivalent).

**Examples**：
```
select k, median(v) from aggr group by k order by k;
+---+-----------+
| K | MEDIAN(V) |
|---+-----------|
| 1 |  20.00000 |
| 2 |  22.50000 |
| 3 |      NULL |
+---+-----------+
```

### Why are the changes needed?
The mainstream database supports `MEDIAN` show below:
**Snowflake**
https://docs.snowflake.com/en/sql-reference/functions/median.html

**Oracle**
https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/MEDIAN.html#GUID-DE15705A-AC18-4416-8487-B9E1D70CE01A

**ClickHouse**
https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/median

**Redshift**
https://docs.aws.amazon.com/redshift/latest/dg/r_MEDIAN.html

**Teradata**
https://docs.teradata.com/r/Teradata-VantageTM-SQL-Functions-Expressions-and-Predicates/March-2019/Ordered-Analytical/Window-Aggregate-Functions/MEDIAN

**DB2**
https://www.ibm.com/docs/en/db2/11.5?topic=functions-median

**Vertica**
https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Functions/Analytic/MEDIANAnalytic.htm?tocpath=SQL%20Reference%20Manual%7CSQL%20Functions%7CAnalytic%20Functions%7C_____20

**H2**
http://www.h2database.com/html/functions-aggregate.html#median

**Sybase**
https://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.infocenter.dc01776.1601/doc/html/san1278453109663.html

**Exasol**
https://docs.exasol.com/db/latest/sql_references/functions/alphabeticallistfunctions/median.htm

**Yellowbrick**
https://www.yellowbrick.com/docs/5.2/ybd_sqlref/median.html

**Mariadb**
https://mariadb.com/kb/en/median/

**Singlestore**
https://docs.singlestore.com/db/v7.6/en/reference/sql-reference/aggregate-functions/median.html

**InfluxDB**
https://docs.influxdata.com/flux/v0.x/stdlib/universe/median/

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New tests.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @MaxGekk cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/percentiles.scala [hunk 1]
```diff
@@ -359,6 +359,32 @@ case class Percentile(
...
+// scalastyle:on line.size.limit
+case class Median(child: Expression)
+  extends AggregateFunction
+    with RuntimeReplaceableAggregate
+    with ImplicitCastInputTypes
+    with UnaryLike[Expression] {
+  private lazy val percentile = new Percentile(child, Literal(0.5, DoubleType))
+  override def replacement: Expression = percentile
+  override def nodeName: String = "median"
+  override def inputTypes: Seq[AbstractDataType] = percentile.inputTypes
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/percentiles.scala:383]: are we sure? `Percentile` has 3 inputs but `Median` only has one

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql [hunk 2]
```diff
@@ -273,3 +273,16 @@ SELECT
 FROM aggr
 GROUP BY k
 ORDER BY k;
+
+-- SPARK-39320: Add the MEDIAN() function
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/group-by.sql:277]: nit: can we add a new golden file for these statistical functions?

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/group-by.sql:277]: and we can test both window and group-by there.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/percentiles.scala [hunk 3]
```diff
@@ -359,6 +359,32 @@ case class Percentile(
...
+// scalastyle:on line.size.limit
+case class Median(child: Expression)
+  extends AggregateFunction
+    with RuntimeReplaceableAggregate
+    with ImplicitCastInputTypes
+    with UnaryLike[Expression] {
+  private lazy val percentile = new Percentile(child, Literal(0.5, DoubleType))
+  override def replacement: Expression = percentile
+  override def nodeName: String = "median"
+  override def inputTypes: Seq[AbstractDataType] = percentile.inputTypes.take(1)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/percentiles.scala:383]: isn't it just `Seq(DoubleType)`?

File: sql/core/src/test/resources/sql-tests/inputs/percentiles.sql [hunk 4]
```diff
@@ -0,0 +1,208 @@
+-- Test data.
+CREATE OR REPLACE TEMPORARY VIEW aggr AS SELECT * FROM VALUES
+(0, 0), (0, 10), (0, 20), (0, 30), (0, 40), (1, 10), (1, 20), (2, 10), (2, 20), (2, 25), (2, 30), (3, 60), (4, null)
+AS aggr(k, v);
+CREATE OR REPLACE TEMPORARY VIEW basic_pays AS SELECT * FROM VALUES
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/percentiles.sql:5]: please add a blank line between SQL statements.

File: sql/core/src/test/resources/sql-tests/inputs/percentiles.sql [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+('Pamela Castillo','SCM',11303),
+('Larry Bott','SCM',11798),
+('Barry Jones','SCM',10586)
+AS basic_pays(employee_name, department, salary);
+
+SELECT
+  percentile_cont(0.25) WITHIN GROUP (ORDER BY v),
+  percentile_cont(0.25) WITHIN GROUP (ORDER BY v DESC)
+FROM aggr;
+SELECT
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/percentiles.sql:29]: ditto

**Review by cloud-fan — APPROVED** [review]
LGTM except for some nits.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/percentiles.scala (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/percentiles.scala:383]: It seems we could supports ansi interval too.
`TypeCollection(NumericType, YearMonthIntervalType, DayTimeIntervalType)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/percentiles.scala [hunk 6]
```diff
@@ -359,6 +359,32 @@ case class Percentile(
...
+       5.0
+  """,
+  group = "agg_funcs",
+  since = "3.4.0")
+// scalastyle:on line.size.limit
+case class Median(child: Expression)
+  extends AggregateFunction
+    with RuntimeReplaceableAggregate
+    with ImplicitCastInputTypes
+    with UnaryLike[Expression] {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/percentiles.scala:379]: I got it.

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master.
Thank you, @beliefer and @cloud-fan for review.

**Comment by beliefer** [comment]
@MaxGekk @cloud-fan Thank you for review.

**Pull request closed without merging**